persist/fs/commitlog/reader.go

```diff
@@ -109,58 +165,171 @@ func (r *reader) Read() (
 	annotation ts.Annotation,
 	resultErr error,
 ) {
-	entry, err := r.readEntry()
-	if err != nil {
-		resultErr = err
-		return
+	if r.nextIndex == 0 {
+		go r.readLoop()
+		for i, decoderBuf := range r.decoderBufs {
+			go r.decoderLoop(decoderBuf, r.outBufs[i])
+		}
 	}
+	// Data is written into these channels in round-robin fashion, so reading them
+	// one at a time in the same order results in an ordered stream.
+	rr, ok := <-r.outBufs[r.nextIndex%r.numConc]
+	if !ok {
+		return Series{}, ts.Datapoint{}, xtime.Unit(0), ts.Annotation(nil), io.EOF
+	}
+	r.nextIndex++
+	return rr.series, rr.datapoint, rr.unit, rr.annotation, rr.resultErr
+}
 
-	metadata, ok := r.metadataLookup[entry.Index]
-	if !ok && len(entry.Metadata) != 0 {
-		r.metadataDecoder.Reset(entry.Metadata)
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
-		if err != nil {
-			resultErr = err
+func (r *reader) readLoop() {
+	index := 0
+	eofFound := false
+	for {
+		select {
+		case <-r.cancelCtx.Done():
+			for _, decoderBuf := range r.decoderBufs {
+				close(decoderBuf)
+			}
+			r.shutdownChan <- r.close()
 			return
+		default:
+			if eofFound {
+				continue
+			}
+			data, err := r.readChunk()
+			if err == io.EOF {
+				eofFound = true
```

Comment:
Is `data` valid if `err == io.EOF`?

Reply:
data will be nil, but the code handles that path

Reply:
Feels wrong to bake in that assumption to the code. Why not just terminate if you're done?

Reply:
Basically the iterator that is calling Read() on the Reader is expecting back: 
```
(
	series Series,
	datapoint ts.Datapoint,
	unit xtime.Unit,
	annotation ts.Annotation,
	resultErr error,
)
```
 on every functional call.

In order for me to satisfy that interface, when I encounter an io.EOF I have to propagate the error through the rest of the code so that the iterator can receive it. That means continuing through the code path one more time where I send the data through the channel with these lines:

```
r.decoderBufs[index%r.numConc] <- decoderArg{
				bytes: data,
				err:   err,
			}
```

Does that make sense?

I could copy that code into the if statement and return early immediately if you think that would be cleaner

Reply:
I might be misunderstanding you if you want to talk in person

Reply:
What I mean is, you're using `data == nil` and/or `err == io.EOF` as sentinels indicating the decoders should terminate. Would be cleaner to use another way to indicate termination to the decoders, e.g. a closeCh channel per decoder, or a field in the decoderArg

Reply:
Hmm I'm still not quite following you. The decoders are closed by an explicit call to shutdown(), they're not waiting to see if err == io.EOF, they just pass those values along.

Let's sync in person on this, I might still be misunderstanding